File: src/converter/plugins/convertVariableDeclaration.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {NamespaceInfoService, namespaceInfoServiceKey} from "./NamespaceInfoPlugin.js";
import {CommentService, commentServiceKey} from "./CommentPlugin.js";
import {ConfigurationService, configurationServiceKey} from "./ConfigurationPlugin.js";

export const convertVariableDeclaration = createSimplePlugin((node, context, render) => {
    if (!ts.isVariableDeclaration(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    // skip initializer
    node.initializer && checkCoverageService?.cover(node.initializer)

    const commentService = context.lookupService<CommentService>(commentServiceKey)
    const configurationService = context.lookupService<ConfigurationService>(configurationServiceKey)
    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    const namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)

    const modifier = node.parent.flags & ts.NodeFlags.Const
        ? "val "
        : "var "

    const name = render(node.name)

    const namespace = typeScriptService?.findClosest(node, ts.isModuleDeclaration)

    let externalModifier = "external "

    if (namespace !== undefined && namespaceInfoService?.resolveNamespaceStrategy(namespace) === "object") {
        externalModifier = ""
    }

    let leadingComment = ""

    if (
        configurationService?.configuration.granularity === "top-level"
        && (
            namespace === undefined
            || namespaceInfoService?.resolveNamespaceStrategy(namespace) === "package"
        )
    ) {
        leadingComment = commentService?.renderLeadingComments(node.parent) ?? ""
    }

    let type: string

    if (node.type) {
        type = render(node.type)
    } else {
        type = "Any? /* should be inferred */" // TODO: infer types
    }

    return `${leadingComment}${externalModifier}${modifier}${name}: ${type}`
})

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertVariableDeclaration.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.configuration.Granularity
import io.github.sgrishchenko.karakum.configuration.NamespaceStrategy
import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node
import typescript.NodeFlags
import typescript.isVariableDeclaration

val convertVariableDeclaration = createSimplePlugin plugin@{ node: Node, context, render ->
    if (!isVariableDeclaration(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    // skip initializer
    node.initializer?.let { checkCoverageService?.cover(it) }

    val commentService = context.lookupService<CommentService>(commentServiceKey)
    val configurationService = context.lookupService<ConfigurationService>(configurationServiceKey)
    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
    val namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)

    val modifier = if (node.parent.flags.toString().toInt() and NodeFlags.Const.toString().toInt() != 0) {
        "val "
    } else {
        "var "
    }

    val name = render(node.name)

    val namespace = typeScriptService?.findClosestNamespace(node)

    var externalModifier = "external "

    if (namespace != null && namespaceInfoService?.resolveNamespaceStrategy(namespace) == NamespaceStrategy.`object`) {
        externalModifier = ""
    }

    var leadingComment = ""

    if (
        configurationService?.configuration?.granularity === Granularity.topLevel
        && (
            namespace == null
            || namespaceInfoService?.resolveNamespaceStrategy(namespace) == NamespaceStrategy.`package`
        )
    ) {
        leadingComment = commentService?.renderLeadingComments(node.parent) ?: ""
    }

    val type = node.type
        ?.let { render(it) }
        ?: "Any? /* should be inferred */" // TODO: infer types

    "${leadingComment}${externalModifier}${modifier}${name}: $type"
}
